applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.cpp

```diff
@@ -65,20 +68,24 @@ namespace Kratos
 
         rResponseGradient.clear();
 
-        if( rAdjointElement.Id() == mpNeighboringElement->Id() )
-        {
-            DofsVectorType dofs_of_element;
-            mpNeighboringElement->GetDofList(dofs_of_element, rProcessInfo);
-
-            const DoubleVariableType& r_traced_adjoint_dof =
-                KratosComponents<DoubleVariableType>::Get(std::string("ADJOINT_") + mTracedDofLabel);
-
-            for(IndexType i = 0; i < dofs_of_element.size(); ++i)
-            {
-                if (dofs_of_element[i]->Id() == mpTracedNode->Id() &&
-                    dofs_of_element[i]->GetVariable() == r_traced_adjoint_dof)
-                {
-                    rResponseGradient[i] = -1;
+        std::map<std::pair<IndexType, IndexType>, IndexType>::iterator it;
+        for (it = mElementNodeMap.begin(); it != mElementNodeMap.end(); ++it) {
+            const auto element_id = (it->first).first;
+            const auto node_id = it->second;
+
+            if( rAdjointElement.Id() == element_id ) {
+                const Variable<double>* adjoint_solution_variable = &KratosComponents<Variable<double>>::Get("ADJOINT_" + mTracedDofLabel + "_X");
```

Comment:
this should be done outside the loop.

Reply:
Also, why "_X" ?

Reply:
I use _X to find later on the index of the first entry where I have to assemble the adjoint load

Reply:
got it.